### ISSUE TITLE
Add support for force renewal

### ISSUE THREAD

**Issue by jihu** [PKISharp/ACME-PS#89]
It would be very useful if this tool supported a forced renewal of an existing certificate.

### Use case:
Scenario: Somehow the current certificate on the server gets broken, maybe from someone working in the prod environment but thinking he is in a test/stage environment, or just importing an Azure template for an Application Gateway and it overwrites the valid certificate with an old one. There are several different ways this can happen.

If this happens in a production environment, one wants to correct the problem as quickly as possible. At that moment, it might not be feasible to resolve the problem by uploading the latest correctly generated certificate, because one might not have easy access to it. And generating a new one manually also might take too long. _Remember, the website is down, for all intents and purposes, so one needs a super quick solution._ Think a _click-of-a-button-quick_.

I think that the quickest and most straight forward way to solve this were to trigger the regular renewal process, but with a parameter that forces the renewal. Otherwise the CA (like Letsencrypt) might ignore the renewal request since the existing certificate (according to their system) is not old enough. Letsencrypt supports forcing a renewal, and I can only assume that the other ones does so too, so it makes sense that the ACME-PS scripts supports this too.

Of course it is not recommended to force the renewal every time, but one could create a separate runbook trigger for this, that only is triggered manually in the Azure Portal, or using a Webhook.

**Comment by glatzert** [comment]
By any chance, do you have references to the let's encrypt behavior? Then I'll probably be able to implement it.

**Comment by glatzert** [comment]
The only references I can find to a "forced renewal" are for certbot.
ACME-PS does not enforce any renewal policy - you are free to renew, whenever you like to.

I cannot find any references other than certbot, so I'll need your help to identify work which needs to be done to help your usecase.

**Comment by jihu** [comment]
Hi @glatzert ,
Sorry for late reply. I have no idea how to find this. I just found many pages online that talk about letsencrypt and forcing a renewal, so I just assumed that such a "parameter/setting" was supported by letsencrypt itself.

The reason I think that is is needed is that when we currently run the cert renewal (using the Azure runbook code that uses ACME-PS) it doesn't renew the cert when running too recently after the creation. But we would like a way to force the renewal, even if the cert was created just minutes ago.

**Comment by glatzert** [comment]
I cannot find any references for "renwal" in RFC 8555 and I'm not proficient in python, so I cannot read the certbot code to look, what happens there, so I currently cannot implement the use-case.
Nevertheless, I'd implement it, if I can find someone who's able to explain me how the force_renew switch would change the ACME-request

**Comment by webprofusion-chrisc** [comment]
Let's Encrypt doesn't care if you ask for the same certificate again (renew) as long as you don't ask for the same certificate (same set of domains) more than 5 times per week (https://letsencrypt.org/docs/rate-limits/) so the normal practice would to just run your acme client again. 

@glazert it won't change the ACME request at all because it's a new order (not really a 'renewal' or anything special), the key thing is just ignoring when it was last requested and proceeding with the request as normal. In certbot this is done with the `--force-renewal` command line flag. Note that you also may need this if Let's Encrypt have done a mass-revoke of certs.

**Comment by glatzert** [comment]
As I thought, since ACME does not mention it at all.

@jihu you should be able to just use the "normal" renewal process. Meaning just issue a new order containing the same identifiers should do the job.

**Comment by jihu** [comment]
@webprofusion-chrisc 

> the key thing is just ignoring when it was last requested and proceeding with the request as normal. 

Could you clarify **who** needs to do this? It sounds like if noone does anything, the renewal will not go through if the last request was done too recently. But I have no idea what code is doing this. Are you saying that this is not part of the ACME-PS code?
Maybe it is the code that uses the ACME-PS?
We use this code:
https://github.com/intelequia/letsencrypt-aw/blob/master/letsencryptaw.ps1

If it would be anyway possible, could you maybe take a quick look and see if that code does something that is stopping us from renewing a cert even though it is was just recently created/renewed?

**Comment by webprofusion-chrisc** [comment]
@jihu I don't know that much about acme-ps (I'm just interested, I develop https://certifytheweb.com), but I don't see anything in that code checking how recently the last cert was requested and quitting. I'd suggest that whatever calls that script is probably running on a schedule and if you want to refresh the cert you need to force that script to run. So what calls this script?

**Comment by glatzert** [comment]
Well essentially your script need's to ignore existing certificates.
The Script you refere to is using ACMEv1(/!\ https://community.letsencrypt.org/t/end-of-life-plan-for-acmev1/88430) and the predecessor of this module from @ebekker .

@JohannesHoppe has provided sample code on how to get certificates with ACME-PS on Azure Runbooks in the sample folder. You might want to adapt that code to get an ACMEv2 compatible script.

**Comment by JohannesHoppe** [comment]
This one: https://github.com/PKISharp/ACME-PS/blob/master/samples/AzureRunbookExample.ps1

😉

**Comment by jihu** [comment]
Thanks @glatzert and @JohannesHoppe ! I will ask our Azure specialist to look into that. Although, the output of our current script mentions acme v02 in the run output:

```
ResourceUrl : https://acme-v02.api.letsencrypt.org/directory
NewAccount  : https://acme-v02.api.letsencrypt.org/acme/new-acct
NewAuthz    : 
NewNonce    : https://acme-v02.api.letsencrypt.org/acme/new-nonce
NewOrder    : https://acme-v02.api.letsencrypt.org/acme/new-order
KeyChange   : https://acme-v02.api.letsencrypt.org/acme/key-change
RevokeCert  : https://acme-v02.api.letsencrypt.org/acme/revoke-cert
Meta        : AcmeDirectoryMeta
```
Also, the ACME-PS Module installed is version 1.2.3, according to the Azure portal.

https://www.powershellgallery.com/packages/ACME-PS/1.2.3

`PowerShell client module for the ACME protocol Version 2, which can be used to interoperate with the Let's Encrypt(TM) projects certificate servers and any other RFC 8555 compliant server.`

So, it seems that it is using the latest ACME version, right? :)

**Comment by glatzert** [comment]
1.2.3 is the latest ACME-PS Version - it's number is not aligned with the ACME-Protocol Version Number (which is v2 and standardized via RFC 8555)

**Comment by jihu** [comment]
@glatzert Yes, I know. And as the quoted text says, version 1.2.3 of ACME-PS uses ACME version 2.

**Comment by jihu** [comment]
I just noticed that I had copy pasted an incorrect URL above. I just got confirmation that we are using this one:

https://github.com/intelequia/letsencrypt-aw/blob/master/letsencryptaw_v2.ps1

So we are definitely using ACME v2.

I also know now what was causing the confusion for me, in that I thought that the certificate wasn't renewed. In fact, the certificate **WAS** renewed, at least from the perspective of Letsencrypt (and https://crt.sh/ confirmed this). **BUT** the Application Gateway has a 24h cache of the cert (or 24 hour polling interval of the keyvault where the cert is stored). That's why I kept seeing the old cert. I waited at most maybe one hour or so. Up to 24 hour delay of the new cert is crazy, if you ask me. 30 minutes tops, should be reasonable, and I would prefer much less than that. But at least now we know the reason.
